Source/Dafny/Translator.cs

```diff
@@ -8020,16 +8019,126 @@ private class WFOptions
-
-    public static string Requires(int arity) {
-      return "Requires" + arity;
-    }
-
-    public static string Reads(int arity) {
-      return "Reads" + arity;
+    // <summary>
+    // A representation for a function handle's `apply`, `requires`, and `reads` functions.
+    // Each of these functions has type 
+    // HandleType -> [Heap, Box, ..., Box] -> X
+    // where `X` is
+    // <list type="bullet">
+    //   <item> `Box` for `apply` </item>
+    //   <item> `bool` for `requires` </item>
+    //   <item> `Set Box` for `reads` </item>
+    // </list>
+    // </summary>
+    public enum ReqReadsApp { Requires, Reads, Apply }
+    
+    public static string RraName(ReqReadsApp rra, int arity) {
+      return rra.ToString() + arity;
+    }
+
+    public static ReqReadsApp RraFromString(string rra) {
+      return (ReqReadsApp) Enum.Parse(typeof(ReqReadsApp), char.ToUpper(rra[0]) + rra.Substring(1)); // todo (MR) super ugly
+    }
+
+    // <summary>
+    // The type of the values in a `Requires`/`Reads`/`Apply` map (see type `X` in the description of <see cref="ReqReadsApp"/>)
+    // </summary>
+    public Bpl.Type RraReturnType(ReqReadsApp rra, Bpl.Type defaultRetType = null) {
+      switch (rra) {
+        case ReqReadsApp.Requires:
+          return Bpl.Type.Bool;
+        case ReqReadsApp.Reads:
+          return TrType(new SetType(true, program.BuiltIns.ObjectQ()));
+        case ReqReadsApp.Apply:
+          return defaultRetType ?? predef.BoxType;
+        default:
+          throw new ArgumentException("invalid ReqReadsApp value");
+      }
+    }
+
+    // The type of a `Requires`/`Reads`/`Apply` function applied to a handle, i.e. the
+    // type of the map for each of the above functions:
+    // `[Heap, Box, ..., Box]bool` for `requires`
+    // `[Heap, Box, ..., Box]Set Box` for `reads`
+    // `[Heap, Box, ..., Box]Box` for `apply`
+    private Bpl.Type RraMapType(
+        IToken tok,
+        int arity,
+        ReqReadsApp rra,
+        Bpl.Type handleReturnType = null
+    ) {
+      var mapKeyTypes = from i in Enumerable.Range(0, arity) select predef.BoxType;
+      return new Bpl.MapType(tok, new List<Bpl.TypeVariable>(), Cons(predef.HeapType, mapKeyTypes.ToList()), RraReturnType(rra, handleReturnType));
+    }
+
+    // <summary>
+    // Constructs a Boogie `Requires`/`Reads`/`Apply` function call applied to a handle, e.g. <c>Requires1(f)</c>
+    // </summary>
+    public Bpl.Expr RequiresReadsApplyFunction(
+        IToken tok,
+        int arity,
+        Bpl.Expr handle,
+        ReqReadsApp rra,
+        Bpl.Type returnType = null
+    ) {
+      return FunctionCall(tok, RraName(rra, arity), RraMapType(tok, arity, rra, returnType), handle);
+    }
+
+    // <summary>
+    // Constructs a Boogie `Requires`/`Reads`/`Apply` function call applied to a handle, heap, and function arguments,
+    // e.g. <c>Requires1(f)[heap, x]</c>
+    // </summary>
+    public Bpl.Expr RequiresReadsApplyCall(
+    IToken tok,
```

Comment:
Indent the parameters. Also, note that the indentation level is different on lines 8078, 8092, and 8108.